### PR TITLE
Avoid saving self for`softmax` and `log_softmax`

### LINKED ISSUE #64000: Softmax/logsoftmax use more memory for backward than needed
## 🚀 Feature
Backward formulas for softmax/logsoftmax reference `self`, even though `self` is not required for computation, e.g. 
```
- name: _log_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
  self: _log_softmax_backward_data(grad, result, dim, self)
 ```
Therefore, we retain reference to `self` and use more memory than needed. We should try to avoid that. 
Related: #63027 (more memory than needed used for ReLU), with the same pitch and motivation. 

The explanation for adding `self` to backward/double backward formulas is in #6786, "in softmax_backward, self argument is added just so that gradient of self can be computed in double backward. Self tensor itself is not necessary neither for softmax_backward, nor for softmax_double_backward." There was no solution proposed in #6786. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes https://github.com/pytorch/pytorch/issues/64000
 - updates double backward formula to compute grad wrt output instead of self
 - ~~In some of the error messages, we still refer to the dtype of the input, even though we are now checking the dtype of the output~~

### CODE REVIEW

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -907,13 +907,12 @@ TORCH_IMPL_FUNC(log_softmax_backward_cuda_out) (
...
   int64_t dim,
-  const Tensor& input,
   const Tensor& grad_input) {
-  bool half_to_float = grad.scalar_type() != input.scalar_type();
+  bool half_to_float = grad.scalar_type() != output.scalar_type();
   if (half_to_float) {
     TORCH_CHECK(
         (grad.scalar_type() == ScalarType::Float &&
-         input.scalar_type() == ScalarType::Half),
+         output.scalar_type() == ScalarType::Half),
```
Comment (albanD) [inline aten/src/ATen/native/cuda/SoftMax.cu:915]: some error messages are not updated here and below right?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -1234,26 +1234,12 @@ Tensor log_sigmoid_double_backward(const Tensor & grad, const Tensor & input) {
...
-  auto gI_t3 = ggI_out_sum_output * gO_out_sum;
-  return gI_t0 - gI_t1 - gI_t2 + gI_t3;
+Tensor softmax_double_backward(const Tensor& grad, const Tensor& grad_output, int dim, const Tensor& output) {
+  return grad_output * grad - (output * grad_output).sum(dim, true) * grad - grad_output * (output * grad).sum(dim, true);
 }
 
 Tensor log_softmax_double_backward(const Tensor & grad, const Tensor & grad_output, int dim, const Tensor & output) {
-  auto z = output.exp();
-  return z * grad_output.sum(dim, true) * ((grad * z).sum(dim, true) - grad);
+  return -grad_output.sum(dim, true) * output.exp() * grad;
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1242]: this can be moved back inline in derivatives.yaml, same for the one above maybe?

**Review by albanD — COMMENTED** [review]
I didn't check the formula in detail. Can we trust CI to catch an error there if there is one?

File: aten/src/ATen/native/SoftMax.cpp [hunk 3]
```diff
@@ -65,7 +64,7 @@ TORCH_META_FUNC(_softmax_backward_data)
   auto grad_input_options =
       grad.options().memory_format(LEGACY_CONTIGUOUS_MEMORY_FORMAT);
 
-  bool half_to_float = grad.scalar_type() != input.scalar_type();
+  bool half_to_float = grad.scalar_type() != output.scalar_type();
```
Comment (ngimel) [inline aten/src/ATen/native/SoftMax.cpp:67]: hm, no, in the half-> float  case `output` will be float, `grad` will be float too, but you'd need to produce `half` gradInput.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:1242]: I changed this one, but the formula for `softmax_double_backward` might be slightly more tricky because `grad.to(output.dtype())` needs to be repeated a couple times.

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -1939,9 +1939,9 @@
   grad_output: log_sigmoid_backward(grad, self, buffer)
   self: log_sigmoid_double_backward(grad * grad_output, self)
 
-- name: _log_softmax_backward_data(Tensor grad_output, Tensor output, int dim) -> Tensor
+- name: _log_softmax_backward_data(Tensor grad_output, Tensor output, int dim, ScalarType input_dtype) -> Tensor
   grad_output: grad.to(output.dtype()) - (grad.to(output.dtype()) * output.exp()).sum(dim, true)
-  output: log_softmax_double_backward(grad.to(output.dtype()), grad_output, dim, output).to(output.dtype())
+  output: log_softmax_double_backward(grad.to(output.dtype()), grad_output, dim, output).to(input_dtype)
```
Comment (ngimel) [inline tools/autograd/derivatives.yaml]: should it be `output.dtype()` here, `output` grad should have the same dtype as `output`?

Comment (soulitzer) [reply tools/autograd/derivatives.yaml]: Whoops, good catch

**Comment by soulitzer** [comment]
> I didn't check the formula in detail. Can we trust CI to catch an error there if there is one?

gradgradcheck is indeed enabled for `(log)_softmax`'s OpInfo, so we should be good

**Comment by ngimel** [comment]
This lgtm, did we check that `input` is indeed not saved now?

**Comment by soulitzer** [comment]
Yeah, see below paste of their generated VariableType kernels:

<details>
<summary>
`_softmax` (after)
</summary>

```
at::Tensor _softmax(c10::DispatchKeySet ks, const at::Tensor & self, int64_t dim, bool half_to_float) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  
  (void)_any_requires_grad;
  std::shared_ptr<SoftmaxBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<SoftmaxBackward0>(new SoftmaxBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_scalar_type = self.scalar_type();
    grad_fn->dim = dim;
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_softmax(ks & c10::after_autograd_keyset, self_, dim, half_to_float);
  })();
  auto result = std::move(_tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  if (result.has_storage()) AT_ASSERT(result.storage().use_count() == 1, "function: _softmax");
  AT_ASSERT(result.use_count() <= 1, "function: _softmax");
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  throw_error_for_complex_autograd(result, "_softmax");
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self)), "Trying to use forward AD with _softmax that does not support it.");
  if (grad_fn) {
    grad_fn->result_ = SavedVariable(result, true);
  }
  return result;
}
```

</details>

<details>
<summary>
`_log_softmax` (after)
</summary>

```
at::Tensor _log_softmax(c10::DispatchKeySet ks, const at::Tensor & self, int64_t dim, bool half_to_float) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  
  (void)_any_requires_grad;
  std::shared_ptr<LogSoftmaxBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<LogSoftmaxBackward0>(new LogSoftmaxBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_scalar_type = self.scalar_type();
    grad_fn->dim = dim;
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_log_softmax(ks & c10::after_autograd_keyset, self_, dim, half_to_float);
  })();
  auto result = std::move(_tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  if (result.has_storage()) AT_ASSERT(result.storage().use_count() == 1, "function: _log_softmax");
  AT_ASSERT(result.use_count() <= 1, "function: _log_softmax");
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  throw_error_for_complex_autograd(result, "_log_softmax");
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self)), "Trying to use forward AD with _log_softmax that does not support it.");
  if (grad_fn) {
    grad_fn->result_ = SavedVariable(result, true);
  }
  return result;
}
```

</details>

<details>
<summary>
`_softmax` (before)
</summary>

```
at::Tensor _softmax(c10::DispatchKeySet ks, const at::Tensor & self, int64_t dim, bool half_to_float) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  
  (void)_any_requires_grad;
  std::shared_ptr<SoftmaxBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<SoftmaxBackward0>(new SoftmaxBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
    grad_fn->dim = dim;
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_softmax(ks & c10::after_autograd_keyset, self_, dim, half_to_float);
  })();
  auto result = std::move(_tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  if (result.has_storage()) AT_ASSERT(result.storage().use_count() == 1, "function: _softmax");
  AT_ASSERT(result.use_count() <= 1, "function: _softmax");
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  throw_error_for_complex_autograd(result, "_softmax");
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self)), "Trying to use forward AD with _softmax that does not support it.");
  if (grad_fn) {
    grad_fn->result_ = SavedVariable(result, true);
  }
  return result;
}
```

</details>

<details>
<summary>
`_log_softmax` (before)
</summary>

```
at::Tensor _log_softmax(c10::DispatchKeySet ks, const at::Tensor & self, int64_t dim, bool half_to_float) {
  auto& self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );
  
  (void)_any_requires_grad;
  std::shared_ptr<LogSoftmaxBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<LogSoftmaxBackward0>(new LogSoftmaxBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_ = SavedVariable(self, false);
    grad_fn->dim = dim;
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_log_softmax(ks & c10::after_autograd_keyset, self_, dim, half_to_float);
  })();
  auto result = std::move(_tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  if (result.has_storage()) AT_ASSERT(result.storage().use_count() == 1, "function: _log_softmax");
  AT_ASSERT(result.use_count() <= 1, "function: _log_softmax");
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  throw_error_for_complex_autograd(result, "_log_softmax");
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self)), "Trying to use forward AD with _log_softmax that does not support it.");
  if (grad_fn) {
    grad_fn->result_ = SavedVariable(result, true);
  }
  return result;
}
```

</details>

**Comment by soulitzer** [comment]
Made a PR to fix failing XLA build here: https://github.com/pytorch/xla/pull/3137

**Comment by soulitzer** [comment]
Tests are all passing except XLA now (which should ready to merge any time). This is ready for final review!

**Review by albanD — APPROVED** [review]
LGTM

**Comment by suo** [comment]
It appears that this diff broke master: https://hud.pytorch.org/commit/fb412bdd80a04e648492dd26c3400a87e8490198

Sadly the CI did not run the affected jobs on PR (cc @zhouzhuojie, @malfet) but I will have to revert.

**Comment by malfet** [comment]
All CI history has been lost on PR after it was re-opened.
@soulitzer did you force-push changes to it?

**Comment by soulitzer** [comment]
@malfet Yes, I did force-push changes. Is that to be avoided?
I still see past runs for this PR on circle-ci here: https://app.circleci.com/pipelines/github/pytorch/pytorch?branch=pull%2F65242

**Comment by suo** [comment]
@soulitzer I recommend using ghstack and ghimport in your workflow, it will preserve all history nicely.
